sympy/stats/stochastic_process_types.py

```diff
@@ -759,3 +767,208 @@ def limiting_distribution(self):
+
+    @property
+    def distribution(self):
+        return self.args[2]
+
+    def __new__(cls, sym, p):
+        if p > 1 or p < 0:
+            raise ValueError("Probability must be in between 0 and 1")
+        p = sympify(p)
+        sym = _symbol_converter(sym)
+        state_space = _set_converter([0,1])
+        return Basic.__new__(cls, sym, state_space, BernoulliDistribution(p), p)
+
+    def probability_of_rth_success(self, ntrial, rsuccess):
+        """
+        Computes the probability of `rth` success in `nth` trail.
+
+        Parameters
+        ==========
+
+        ntrial: Positive Integer (ntrial > 1),
+          Represents the `nth` trial.
+        rsuccess: Positive Integer (rsuccess >= 1),
+          Represents `rth` success event (rsuccess <= ntrial).
+
+        Returns
+        =======
+
+        Probability that `rth` success occurs on `nth` trial.
+
+        """
+        if S(rsuccess).is_Integer and S(ntrial).is_Integer:
+            if rsuccess <= ntrial and rsuccess >= 1 and ntrial > 1:
+                return (binomial(ntrial - 1, rsuccess - 1) * self.p**rsuccess
+                        * (1 - self.p)**(ntrial - rsuccess))
+            else:
+                raise ValueError("0 =< rth-success <= nth-trial, nth-trial > 1 and ")
+        else:
+            raise ValueError("nth-trial and rth-success must be Positive Integers.")
+
+
+    def probability_r_success(self, ntrials, rsuccess, evaluate=True):
+        """
+        Computes the probability of `r` success events among the `n` trials.
+
+        Parameters
+        ==========
+
+        ntrials: Positive Integer,
+          Represents number of trials.
+        rsuccess: Positive Integer including 0, by default is None
+          Represents total number of success events (rsuccess <= ntrials).
+        evaluate: Boolean, by default is False
+          Computes the result numerically.
+
+        Returns
+        =======
+        Proability of getting `r` success in `n` trials.
+            If evaluate is True
+        Distribution of getting `r` success in `n` trials.
+            If evaluate is False
+
+        """
+        if evaluate:
+            if S(ntrials).is_Integer and S(rsuccess).is_Integer:
+                if rsuccess <= ntrials and rsuccess >= 0:
+                    return (binomial(ntrials, rsuccess) * self.p**rsuccess
+                            * (1 - self.p)**(ntrials - rsuccess))
+                else:
+                    raise ValueError("rsuccess <= ntrials and  ntrials, \
+                                     rsuccess must be positive Integers.")
+            else:
+                raise ValueError("ntrials and rsuccess must be postive Integers.")
+        else:
+            if S(ntrials).is_Integer and ntrials > 0:
+                r = symbols('r', natural=True)
+                pmf = binomial(ntrials, r) * (self.p**r) * (1 - self.p)**(ntrials - r)
+                return DiscreteDistributionHandmade(pmf, set = S.Naturals)
```

Comment:
First check for errors then start doing the computation. See other parts of code to observe the pattern. The code becomes confusing if not done in this way.

Reply:
```
if not (S(ntrials).is_Integer and S(rsuccess).is_Integer):
      raise ValueError("ntrials and rsuccess must be postive Integers.")
return (binomial(ntrials, rsuccess) * self.p**rsuccess
                            * (1 - self.p)**(ntrials - rsuccess))
```
Should I change this way?

Reply:
Yes that looks better. Filtering should happen at the entry point and then the computation.

Reply:
Yes, will Change that.